File: 151_reverse_words_in_a_string/Solution.go
package main

import (
    "fmt"
    "strings"
)

func reverseWords(s string) string {
    s = strings.Trim(s, " ")
    l , r:= len(s) - 1, len(s) - 1
    var result []string
    for l >= 0 {
        for l >= 0 && s[l] != ' ' { l-- }
        result = append(result, s[l + 1: r + 1])
        for l >= 0 && s[l] == ' ' { l-- }
        r = l
    }

    return strings.Join(result, " ")
}

func main() {
    s := "Stay hungry, stay foolish."
    fmt.Println(reverseWords(s))
}

File: 151_reverse_words_in_a_string/Solution.java
class Solution {
    public String reverseWords(String s) {
        s = s.trim();
        int sLen = s.length();
        if (sLen <= 1) return s;
        StringBuilder result = new StringBuilder();
        int i = sLen - 1, j = i;
        while (i >= 0) {
            while (i >= 0 && s.charAt(i) != ' ') i--;
            result.append(s.substring(i + 1, j + 1) + " ");
            while (i >= 0 && s.charAt(i) == ' ') i--;
            j = i;
        }
        return result.toString().trim();
    }

    public static void main(String[] args) {
        String input = "Stay hungry, stay foolish.";
        System.out.println(new Solution().reverseWords(input));
    }
}
